activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, options)
+      @model = model
+      @inserts = inserts
+      @options = options
+    end
+
+    def execute
+      return ActiveRecord::Result.new([], []) if inserts.empty?
+
+      connection.exec_query to_sql, "Bulk Insert"
+    end
+
+    def to_sql
+      raise ArgumentError, "inserts is empty" if inserts.empty?
+
+      connection.build_insert_all_sql(
+        table_name,
+        keys,
+        values_list,
+        updatable_columns: updatable_columns,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        conflict_target: conflict_target)
+    end
+
+    private
+      attr_reader :model, :inserts, :options
+
+      delegate :connection,
+               :table_name,
+               to: :model
+
+      delegate :supports_insert_returning?,
+               :supports_insert_conflict_target?,
+               :supports_insert_on_duplicate_skip?,
+               :supports_insert_on_duplicate_update?,
+               to: :connection
+
+      def primary_keys
```

Comment:
Present methods in order of appearance, or Table of Contents, style. So that means that helper methods like this should be appear below the executing methods that depend on them.